File: python/quick.py
def quick_sort(arr):
    if len(arr) <= 1:
        return arr

    pivot = arr[-1]
    print(f"\nPivot = {pivot}")

    left = []
    right = []

    for x in arr[:-1]:
        print(f"Bandingkan {x} dengan {pivot}", end=" ")
        if x <= pivot:
            left.append(x)
            print("-> Masuk kiri")
        else:
            right.append(x)
            print("-> Masuk kanan")

    print("Kiri :", left)
    print("Kanan:", right)

    return quick_sort(left) + [pivot] + quick_sort(right)


# Input
n = int(input("Masukkan jumlah data: "))
data = list(map(int, input("Masukkan data: ").split()))

# Proses
hasil = quick_sort(data)

# Output
print("\nHasil akhir:", hasil)

File: c++/quick.cpp
#include <iostream>
using namespace std;

void tampil(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

int partition(int arr[], int low, int high, int n) {
    int pivot = arr[high];
    cout << "\nGunakan pivot = " << pivot << endl;

    int i = low - 1;

    for (int j = low; j < high; j++) {
        cout << "Bandingkan " << arr[j] << " <= " << pivot;

        if (arr[j] <= pivot) {
            i++;
            swap(arr[i], arr[j]);
            cout << " -> Tukar\n";
        } else {
            cout << " -> Tidak\n";
        }

        tampil(arr, n);
    }

    swap(arr[i + 1], arr[high]);
    cout << "Tukar pivot\n";
    tampil(arr, n);

    return i + 1;
}

void quickSort(int arr[], int low, int high, int n) {
    if (low < high) {
        int pi = partition(arr, low, high, n);

        cout << "Pivot di posisi " << pi << endl;

        quickSort(arr, low, pi - 1, n);
        quickSort(arr, pi + 1, high, n);
    }
}

int main() {
    int n;
    cout << "Masukkan jumlah data: ";
    cin >> n;

    int arr[100];
    cout << "Masukkan data: ";
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    quickSort(arr, 0, n - 1, n);

    cout << "\nHasil akhir: ";
    tampil(arr, n);

    return 0;
}
